<h1>SQL анализ базы данных книжного сервиса<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Постановка-задачи." data-toc-modified-id="Постановка-задачи.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Постановка задачи.</a></span></li><li><span><a href="#Создание-подключения." data-toc-modified-id="Создание-подключения.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Создание подключения.</a></span></li><li><span><a href="#Описание-данных." data-toc-modified-id="Описание-данных.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Описание данных.</a></span></li><li><span><a href="#Задачи." data-toc-modified-id="Задачи.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задачи.</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года;-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года;</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку." data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку.-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку.</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры." data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры.-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками." data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками.-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Вывод." data-toc-modified-id="Вывод.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод.</a></span></li></ul></div>

## Постановка задачи.

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Создание подключения.

In [1]:
# Импортируем библиотеки.

import pandas as pd 
from sqlalchemy import create_engine 

In [2]:
# Устанавливаем параметры.

db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
 db_config['pwd'], 
 db_config['host'], 
 db_config['port'], 
 db_config['db']) 

In [3]:
# Сохраняем коннектор.
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Описание данных.

Исследуем таблицы — выведем первые строки.

In [4]:
pd.io.sql.read_sql("SELECT * FROM books LIMIT 10;", con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

Таблица books

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

In [5]:
pd.io.sql.read_sql("SELECT * FROM authors LIMIT 10;", con = engine)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.


In [6]:
pd.io.sql.read_sql("SELECT * FROM publishers LIMIT 10;", con = engine)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

In [7]:
pd.io.sql.read_sql("SELECT * FROM ratings LIMIT 10;", con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

In [8]:
pd.io.sql.read_sql("SELECT * FROM reviews LIMIT 10;", con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...


**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Задачи.

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [9]:
pd.io.sql.read_sql("SELECT * FROM books LIMIT 10;", con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [10]:
pd.io.sql.read_sql(
    "SELECT COUNT(book_id) FROM books WHERE publication_date > '2000-01-01';", con = engine) 

count
0    819

После 1 января 2000 года вышло 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [11]:
pd.io.sql.read_sql(
    '''SELECT b.title, 
    COUNT(DISTINCT r.review_id), 
    AVG(rat.rating) 
    FROM reviews AS r 
    LEFT JOIN books AS b ON b.book_id = r.book_id 
    LEFT JOIN ratings AS rat ON b.book_id = rat.book_id 
    GROUP BY b.title
    ORDER BY COUNT(DISTINCT r.review_id) DESC;''', con = engine)

title  count       avg
0                                  Memoirs of a Geisha      8  4.138462
1                              Twilight (Twilight  #1)      7  3.662500
2               The Da Vinci Code (Robert Langdon  #2)      6  3.830508
3                                             The Road      6  3.772727
4    Harry Potter and the Prisoner of Azkaban (Harr...      6  4.414634
..                                                 ...    ...       ...
988                  Naked Empire (Sword of Truth  #8)      1  3.500000
989                                Moo  Baa  La La La!      1  3.000000
990                Merrick (The Vampire Chronicles #7)      1  4.000000
991                                          Babyville      1  3.500000
992                                         Lysistrata      1  4.000000

[993 rows x 3 columns]

In [12]:
pd.io.sql.read_sql('''
SELECT 
    b.title AS title,
    b.book_id AS id,
    AVG(r.rating) AS avg_rating,
    COUNT(DISTINCT rev.text)
FROM
    books AS b
INNER JOIN ratings AS r ON r.book_id = b.book_id
INNER JOIN reviews AS rev ON rev.book_id = b.book_id
GROUP BY
     b.title,
     b.book_id
ORDER BY
      COUNT(DISTINCT rev.text) DESC
limit 20
''', con = engine)

title   id  avg_rating  count
0                             Twilight (Twilight  #1)  948    3.662500      7
1                                       The Alchemist  627    3.789474      6
2                           Outlander (Outlander  #1)  497    4.125000      6
3                 The Hobbit  or There and Back Again  750    4.125000      6
4                                     Eat  Pray  Love  207    3.395833      6
5                                      The Book Thief  656    4.264151      6
6   The Curious Incident of the Dog in the Night-Time  695    4.081081      6
7                              The Catcher in the Rye  673    3.825581      6
8                           The Giver (The Giver  #1)  733    3.750000      6
9                                    The Glass Castle  734    4.206897      6
10  The Lightning Thief (Percy Jackson and the Oly...  779    4.080645      6
11                                           The Road  854    3.772727      6
12                                Water for Elephants  963    3.977273      6
13             The Da Vinci Code (Robert Langdon  #2)  696    3.830508      6
14  Harry Potter and the Chamber of Secrets (Harry...  299    4.287500      6
15  Harry Potter and the Prisoner of Azkaban (Harr...  302    4.414634      6
16     A Clash of Kings  (A Song of Ice and Fire  #2)   15    4.391304      5
17                Extremely Loud and Incredibly Close  232    3.416667      5
18  Freakonomics: A Rogue Economist Explores the H...  267    3.909091      5
19                                            Dracula  194    3.296296      5

Есть ряд книг с единичными обзорами. Наибольшее количество обзоров - 7. Средняя оценка 3,66.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [13]:
pd.io.sql.read_sql('''SELECT p.publisher,
COUNT(b.title) 
FROM books AS b 
LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id 
WHERE num_pages > 50 
GROUP BY p.publisher 
ORDER BY COUNT(b.title) DESC
LIMIT 15;''', con = engine)

publisher  count
0               Penguin Books     42
1                     Vintage     31
2    Grand Central Publishing     25
3            Penguin Classics     24
4            Ballantine Books     19
5                      Bantam     19
6                     Berkley     17
7          St. Martin's Press     14
8               Berkley Books     14
9                       Delta     13
10  William Morrow Paperbacks     13
11              HarperCollins     12
12  Little  Brown and Company     12
13           Harper Perennial     12
14                  Tor Books     11

Издательство, которое выпустило наибольшее число книг более 50 страниц - Penguin Books, на втором месте Vintage.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [14]:
pd.io.sql.read_sql('''SELECT
      data.author AS author,
      AVG(data.avg_rating) AS avg_rating
FROM
(SELECT
     a.author AS author,
     b.book_id AS id,
     AVG(r.rating) AS avg_rating,
     COUNT(r.rating) AS count_rating
     FROM authors AS a
     INNER JOIN books AS b ON a.author_id = b.author_id
     INNER JOIN ratings AS r ON r.book_id = b.book_id
     GROUP BY author, id
     HAVING COUNT(r.rating) >= 50) AS data
GROUP BY
    author
ORDER BY
    avg_rating DESC;''', con = engine)

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.283844
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.258446
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                           Dan Brown    3.754540
10                                         Lois Lowry    3.750000
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

Автор с самой высокой средней оценкой книг - J.K. Rowling. Не удивительно, Гарри Поттер уже долгое время на слуху, а в те годы был пик его популряности. 

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [15]:
pd.io.sql.read_sql('''
SELECT ROUND(avg(count)) 
FROM (SELECT COUNT(text)
FROM reviews
FULL OUTER JOIN (SELECT username, COUNT(rating_id) as cnt
FROM ratings
GROUP BY username) AS cnt_rating ON cnt_rating.username = reviews.username
WHERE cnt > 50
GROUP BY reviews.username) as count;''', con = engine)                                

round
0   24.0

Среднее количество обзоров от авторов поставивших более 50 оценок - 24.

## Вывод.

После 1 января 2000 года вышло 819 книг.

Есть ряд книг с единичными обзорами. Наибольшее количество обзоров - 7. Средняя оценка 3,66.

Издательство, которое выпустило наибольшее число книг более 50 страниц - Penguin Books, на втором месте Vintage.

Автор с самой высокой средней оценкой книг - J.K. Rowling.

Среднее количество обзоров от авторов поставивших более 50 оценок - 24.